pkg/controller/clusterinstallation/mattermost.go

```diff
@@ -207,6 +206,87 @@ func (r *ReconcileClusterInstallation) updateMattermostDeployment(mi *mattermost
 		return fmt.Errorf("Unable to find mattermost container in deployment")
 	}
 
+	// Run a single-pod job with the new mattermost image to perform any
+	// database migrations before altering the deployment. If this fails,
+	// we will return and not upgrade the deployment.
+	if update {
+		reqLogger.Info(fmt.Sprintf("Running Mattermost image %s upgrade job check", image))
+
+		updateName := "mattermost-image-update-check"
+		job := &batchv1.Job{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      updateName,
+				Namespace: mi.GetNamespace(),
+			},
+			Spec: batchv1.JobSpec{
+				Template: corev1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: map[string]string{"app": updateName},
+					},
+					Spec: original.Spec.Template.Spec,
+				},
+			},
+		}
+
+		// Override values for job-specific behavior.
+		job.Spec.Template.Spec.RestartPolicy = corev1.RestartPolicyNever
+		for i := range job.Spec.Template.Spec.Containers {
+			job.Spec.Template.Spec.Containers[i].Command = []string{"mattermost", "version"}
+		}
+
+		err := r.client.Create(context.TODO(), job)
+		if err != nil && !k8sErrors.IsAlreadyExists(err) {
+			return err
+		}
+		defer func() {
+			err = r.client.Delete(context.TODO(), job)
+			if err != nil {
+				reqLogger.Error(err, "Unable to cleanup image update check job")
+			}
+		}()
+
+		// Wait up to 60 seconds for the update check job to return successfully.
+		timer := time.NewTimer(60 * time.Second)
```

Comment:
What if the upgrade job takes longer than 60s to run? We've had schema changes in the past that can take minutes (and sometimes hours) to apply when modifying large tables. We avoid making changes to those large tables but I wouldn't be surprised if we had upgrades that took 10-15min to complete in the future

Reply:
Interesting. I had forgotten that that was a possibility.

I think that if we want to wait any longer than 60 seconds then we shouldn't use a timer-based approach. Instead, we should start the job and leave it running while the reconcile loop checks on it over time. I can definitely develop this, but will need to do additional testing.

Do we want to merge this version and then do a follow-up PR, or wait until the next iteration is ready?

Reply:
I think it's fine to merge this for now since it's already better than what we have. I've created a ticket here to follow-up https://mattermost.atlassian.net/browse/MM-16276 that we can talk about next sprint planning